8_bit_pic.py

```diff
@@ -11,102 +37,352 @@
+            case "GOTO":
+                tokens.append(GOTOToken(operands[0]))
+            case "MOVLW":
+                tokens.append(MOVLWToken(operands[0]))
+            case _:
+                raise Exception(f"Unexpected opcode '{opcode}' on line {line_num}.")
+
     def tokenize(self) -> Sequence[vm_types.AssemblerToken]:
-        return []
+        tokens: list[vm_types.AssemblerToken] = []
+
+        lines = self.text.split("\n")
+        for i, line in enumerate(lines):
+            line_num = i + 1
+            line = line.strip()
+            if not line:
+                # Skip empty lines.
+                pass
+            elif line[0] == ";":
+                # Skip comment lines.
+                pass
+            elif m := self.PSEUDOOP_RE.match(line):
+                # Pseudo-op
+                pseudoop = m.group("pseudoop").upper()
+                operands = m.group("operands").split(",")
+                self._process_pseudoop(m.group("name"), pseudoop, operands, line_num)
+            elif m := self.OPCODE_RE.match(line):
+                # Opcode
+                opcode = m.group("opcode").upper()
+                operands = m.group("operands") or ""
+                operands = operands.split(",")
+                self._process_opcode(m.group("label"), opcode, operands, line_num, tokens)
+            else:
+                raise Exception(f"Syntax error on line {line_num}.")
+
+        logger.debug(tokens)
+
+        return tokens
 
     def assemble(self):
+        self.byte_code = bytearray()
+
+        for token in self.tokenize():
+            if code_bytes := token.encode(self):
+                self.byte_code.extend(code_bytes)
+
+        logger.debug(self.byte_code)
+        logger.debug(self.symbol_table)
+
         return self.byte_code
 
     def link(self):
         return self.byte_code
 
     def compile(self):
+        self.assemble()
+        self.link()
         return self.byte_code
 
 
-@dataclass
-class CPUFlags: ...
+def resolve_int_value(assembler: vm_types.GenericAssembler, value: str, mask: int) -> int:
+    """Resolve integer referene"""
+    if value in assembler.symbol_table["symbols"]:
+        # Symbol reference.
+        value = assembler.symbol_table["symbols"][value]
+    elif value in assembler.symbol_table["map"]:
+        # Address reference.
+        value = assembler.symbol_table["map"][value]
+    else:
+        # Literal.
+        pass
+    try:
+        int_value = int(value)
+    except ValueError:
+        # Try to handle hexadecimal literals with "h" suffix.
+        int_value = int(value.lower().rstrip("h"), 16)
+    int_value = int(int_value) & mask
+    return int_value
 
 
 @dataclass
-class CPURegisters: ...
-
+class PICCentralProcessingUnit(vm_types.GenericCentralProcessingUnit):
+    RAM: bytearray
+    INSTRUCTION_MAP: ClassVar[
+        MutableMapping[InstructionCodes, Callable[[Self, int], None]]
+    ] = {}
```

Comment:
This decorator based interface feels clunky.

Reply:
Ya def. working through approaches as I am going ... I did add in a way to register variations of the ins. in my last commit (still decorator based tho) ... which is where I might take this ultimately in the toy vm, maybe 🤔

Reply:
Ya def. working through approaches as I am going ... I did add in a way to register variations of the ins. in my last commit (still decorator based tho) ... which is where I might take this ultimately in the toy vm, maybe 🤔